code/game/machinery/turrets/turret_fsm.dm

```diff
@@ -0,0 +1,79 @@
+/datum/state_machine/turret
+	current_state = /decl/state/turret/idle
+	expected_holder_type = /obj/machinery/turret
+
+
+/decl/state/turret
+	var/ray_color = "#FFFFFFFF" // Turrets have a visual indicator of their current state.
+	var/switched_to_sound = null
+
+/decl/state/turret/apply_to(obj/machinery/turret/turret)
+	turret.ray_color = src.ray_color
+	turret.update_icon()
+	if(switched_to_sound)
+		playsound(turret, switched_to_sound, 40, TRUE)
+
+/decl/state/turret/idle
+	ray_color = "#00FF00FF"
+	switched_to_sound = 'sound/machines/triple_beep.ogg'
+	transitions = list(
+		/decl/state_transition/turret/lost_power,
+		/decl/state_transition/turret/shoot,
+		/decl/state_transition/turret/turn_to_bearing
+		)
+
+/decl/state/turret/turning
+	ray_color = "#FFFF00FF"
+	switched_to_sound = 'sound/machines/quiet_beep.ogg'
+	transitions = list(
+		/decl/state_transition/turret/lost_power,
+		/decl/state_transition/turret/shoot,
+		/decl/state_transition/turret/no_enemies
+		)
+
+/decl/state/turret/engaging
+	ray_color = "#FF0000FF"
+	switched_to_sound = 'sound/machines/buttonbeep.ogg'
+	transitions = list(
+		/decl/state_transition/turret/lost_power,
+		/decl/state_transition/turret/turn_to_bearing,
+		/decl/state_transition/turret/no_enemies
+		)
+
+/decl/state/turret/no_power
+	ray_color = "#00000000" // Makes the beam invisible with #RRGGBBAA, not black.
+	transitions = list(
+		/decl/state_transition/turret/shoot,
+		/decl/state_transition/turret/turn_to_bearing,
+		/decl/state_transition/turret/no_enemies
+		)
+
+/decl/state_transition/turret/turn_to_bearing
+	target = /decl/state/turret/turning
+
+/decl/state_transition/turret/turn_to_bearing/is_open(obj/machinery/turret/turret)
+	return !turret.within_bearing()
+
+
+/decl/state_transition/turret/shoot
+	target = /decl/state/turret/engaging
+
+/decl/state_transition/turret/shoot/is_open(obj/machinery/turret/turret)
+	return turret.target && turret.within_bearing()
+
+
+/decl/state_transition/turret/no_enemies
+	target = /decl/state/turret/idle 
+
+/decl/state_transition/turret/no_enemies/is_open(obj/machinery/turret/turret)
+	return !turret.target // TODO
+
+/decl/state_transition/turret/yes_power
+	target = /decl/state/turret/idle 
+
+
+/decl/state_transition/turret/lost_power
+	target = /decl/state/turret/no_power
+
+/decl/state_transition/turret/lost_power/is_open(obj/machinery/turret/turret)
+	return (turret.stat & NOPOWER) && !turret.internal_cell.check_charge(turret.idle_power_usage * CELLRATE) // Should this be made a proc on turrets?
```

Comment:
Probably should make this some variant of `is_powered()` on the turrets so you could potentially override them down the track for wired or immortal (PoI etc) turrets.